academic_observatory_workflows/workflows/pubmed_telescope.py

```diff
@@ -0,0 +1,1623 @@
+
+        snapshot_table_id = bq_sharded_table_id(
+            self.cloud_workspace.project_id, self.bq_dataset_id, f"{self.bq_table_id}_backup", release.start_date
+        )
+        expiry_date = pendulum.now().add(days=self.snapshot_expiry_days)
+        success = bq_snapshot(src_table_id=self.main_table_id, dst_table_id=snapshot_table_id, expiry_date=expiry_date)
+
+        set_task_state(success, kwargs["ti"].task_id, release=release)
+
+    def download_updatefiles(self, release: PubMedRelease, **kwargs):
+        """
+        Download the updatefiles from PubMed's FTP server for this release.
+
+        Unable to do this in parallel due to limitations of their FTP server.
+        """
+
+        updatefiles_datafiles = [datafile for datafile in release.datafile_list if not datafile.baseline]
+
+        success = download_datafiles_from_ftp_server(
+            datafile_list=updatefiles_datafiles,
+            ftp_server_url=self.ftp_server_url,
+            ftp_port=self.ftp_port,
+            reset_ftp_counter=self.reset_ftp_counter,
+            max_download_retry=self.max_download_retry,
+        )
+
+        set_task_state(success, kwargs["ti"].task_id, release=release)
+
+    def upload_downloaded_updatefiles(self, release: PubMedRelease, **kwargs):
+        """Upload downloaded updatefiles files to GCS."""
+
+        # Grab list of files to upload.
+        datafiles_to_upload = [
+            datafile.download_file_path for datafile in release.datafile_list if not datafile.baseline
+        ]
+
+        success = gcs_upload_files(
+            bucket_name=self.cloud_workspace.download_bucket,
+            file_paths=datafiles_to_upload,
+        )
+
+        set_task_state(success, kwargs["ti"].task_id, release=release)
+
+    def transform_updatefiles(self, release: PubMedRelease, **kwargs):
+        """
+        Transform the *.xml.gz files downloaded from PubMed's FTP server into usable json-like files for BigQuery import.
+
+        This is a multithreaded task that pulls the information from the files using the keys defined in self.entity_list.
+        """
+
+        # Get list of datafiles for the baseline to trasform.
+        datafiles_to_transform = [datafile for datafile in release.datafile_list if not datafile.baseline]
+
+        # Process files in batches so that ProcessPoolExecutor doesn't deplete the system of memory
+        for i, chunk in enumerate(get_chunks(input_list=datafiles_to_transform, chunk_size=self.max_processes)):
+            with ProcessPoolExecutor(max_workers=self.max_processes) as executor:
+                logging.info(f"In chunk {i} and processing files: {chunk}")
+
+                futures = [
+                    executor.submit(
+                        transform_pubmed_xml_file_to_jsonl,
+                        datafile,
+                        [self.entity_list["upsert"], self.entity_list["delete"]],
+                    )
+                    for datafile in chunk
+                ]
```

Comment:
I think this would make it clearer what files are being input and created:
```python
futures = []
for datafile in chunk:
    input_path = datafile.download_file_path
    upsert_path = datafile.transform_file_path(entity_type=self.upsert.type)
    delete_path = datafile.transform_file_path(entity_type=self.delete.type)
    futures.append(executor.submit(transform_pubmed, input_path, upsert_path, delete_path))
```